bpf/bpf_lxc.c

```diff
@@ -173,12 +173,20 @@ static __always_inline int ipv6_l3_from_lxc(struct __ctx_buff *ctx,
 			   orig_dip.p4, *dstID);
 	}
 
-	/* If the packet is in the establishing direction and it's destined
-	 * within the cluster, it must match policy or be dropped. If it's
-	 * bound for the host/outside, perform the CIDR policy check.
+	/* When an endpoint connects to itself via service clusterIP, we need
+	 * to skip the policy enforcement. If we didn't, the user would have to
+	 * define policy rules to allow pods to talk to themselves. We still
+	 * want to execute the conntrack logic so that replies can be correctly
+	 * matched.
 	 */
-	verdict = policy_can_egress6(ctx, tuple, SECLABEL, *dstID,
-				     &policy_match_type, &audited);
+	if (!hairpin_flow)
```

Comment:
Just for consistency with the ingress path, should we instead do..

```
	if (ct_state.loopback)
		goto skip_policy_enforcement;
```

And set the label to jump similarly just before the switch statement below?

This will also address the inconsistent policy verdict output concern.

Reply:
For the egress path it's not clear where to place such a label. We want to skip notifications but still want to create the conntrack entry (or we'll break ingress).

> This will also address the inconsistent policy verdict output concern.

I think that is fixed with the new version.

Reply:
Here's what I had in mind:

```
$ git di
diff --git a/bpf/bpf_lxc.c b/bpf/bpf_lxc.c
index 55d340761a19..3b3738ef70e5 100644
--- a/bpf/bpf_lxc.c
+++ b/bpf/bpf_lxc.c
@@ -568,13 +568,15 @@ skip_service_lookup:
         * want to execute the conntrack logic so that replies can be correctly
         * matched.
         */
-       if (!hairpin_flow)
-               /* If the packet is in the establishing direction and it's destined
-               * within the cluster, it must match policy or be dropped. If it's
-               * bound for the host/outside, perform the CIDR policy check.
-               */
-               verdict = policy_can_egress4(ctx, &tuple, SECLABEL, *dstID,
-                                            &policy_match_type, &audited);
+       if (hairpin_flow)
+               goto skip_policy_enforcement;
+
+       /* If the packet is in the establishing direction and it's destined
+       * within the cluster, it must match policy or be dropped. If it's
+       * bound for the host/outside, perform the CIDR policy check.
+       */
+       verdict = policy_can_egress4(ctx, &tuple, SECLABEL, *dstID,
+                                    &policy_match_type, &audited);

        if (ret != CT_REPLY && ret != CT_RELATED && verdict < 0) {
                send_policy_verdict_notify(ctx, *dstID, tuple.dport,
@@ -583,6 +585,7 @@ skip_service_lookup:
                return verdict;
        }

+skip_policy_enforcement:
        switch (ret) {
        case CT_NEW:
                if (!hairpin_flow)
```

Reply:
In particular, the `send_policy_verdict_notify()` on line 190 could still be triggered by hairpin flows here.

Reply:
> Nevermind, that check requires verdict < 0 which would never be true for the hairpin_flow case.

Yep, that's why I didn't take that route. It might be better just for consistency with the ingress path. I also find it a bit easier to read but I might be biased.

Reply:
I'm not too fussed, my predisposition on these is something like (1) Make it readable, (2) Make it consistent and (3) Minimize code changes so that `git blame` is more likely to point to the oldest relevant commits for when you want to dig through and get context about the changes. In that order.

I happen to like the clear "skip_policy_enforcement" label which documents the intent pretty well but this is all much of a muchness :wink: .